Add e2e test for CSI volume limits

jsafrane opened the pull request:
The test creates as many PVs and pods as the driver reports to support on a single node.

Tested with AWS, it's really slow, but it works. Let's see how GCE works with 128 volumes :-)

/kind feature
/sig storage

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

@msau42 @davidz627

msau42 commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,256 @@
+			framework.ExpectNoError(err, "failed waiting for PVs to be deleted")
+		}()
+
+		// Create <limit> nr. of PVCs and pods. With one pod created above, we are 1 pod over the limit.
```
I think we need more PVCs per pod otherwise we will hit 110 pods per node limit first

jsafrane replied:
Why is it so low???
Anyway, I was not able to even create 128 PVs:

```
I0717 13:28:50.828] STEP: Waiting for 129 PVCs Bound
I0717 13:29:00.642] Jul 17 13:29:00.642: INFO: 30/129 of PVCs are Bound
I0717 13:29:07.375] Jul 17 13:29:07.375: INFO: 37/129 of PVCs are Bound
I0717 13:29:15.760] Jul 17 13:29:15.759: INFO: 43/129 of PVCs are Bound
I0717 13:29:22.415] Jul 17 13:29:22.415: INFO: 52/129 of PVCs are Bound
I0717 13:29:30.550] Jul 17 13:29:30.549: INFO: 65/129 of PVCs are Bound
I0717 13:29:37.465] Jul 17 13:29:37.465: INFO: 72/129 of PVCs are Bound
I0717 13:29:45.482] Jul 17 13:29:45.482: INFO: 80/129 of PVCs are Bound
I0717 13:29:53.365] Jul 17 13:29:53.365: INFO: 87/129 of PVCs are Bound

...

I0717 13:59:11.115] Jul 17 13:59:11.115: INFO: 105/129 of PVCs are Bound
I0717 13:59:11.116] Jul 17 13:59:11.115: INFO: Unexpected error occurred: timed out waiting for the condition
```

msau42 replied:
Are you sure it's not because pods couldn't be scheduled?  PD driver uses delayed binding

davidz627 commented:
/cc @davidz627 @hantaowang

davidz627 commented:
@jsafrane need me to run the GCE version or will you handle it? Or the PR is not ready yet?

msau42 commented:
@davidz627 it's getting run as part of pull-kubernetes-e2e-gce-csi-serial

jsafrane commented:
I checked late binding both with AWS and mock driver, something is very fishy there. Some scheduler predicate allows only 16 unbound PVCs in a single pod to be scheduled, a pod with 17 unbound PVCs is Pending forever. My theory is that GCE and/or Azure predicate don't check `provisioner:` and assume the unbound PVCs are theirs. @bertinatto is looking at it.

bertinatto commented:
/retest

https://github.com/kubernetes/kubernetes/pull/80360 has been merged.

bertinatto commented:
/test pull-kubernetes-e2e-gce-csi-serial

jsafrane commented:
@davidz627 @msau42, so I tried with 128 volumes and a pod with 128 volumes failed to run:

```
[snip] rpc error: code = Internal desc = unknown Attach error: failed when waiting for zonal op: operation operation-1564668114867-58f0eaf2d9379-022fe473-7f224368 failed (LIMIT_EXCEEDED): Exceeded limit 'maximum_persistent_disks' on resource 'e2e-4c57e689bd-930d0-minion-group-00hf'. Limit: 128.0
```

Is there any off-by-one error here? Or is there any other pod using one slot (logging, metrics, ...) running? It would be great if you debug the test on your end.

davidz627 commented:
@jsafrane I will look into it

davidz627 commented:
@jsafrane @msau42 yep we have an off-by-one error. The nodes on GCE have a Boot Disk which is counted as an "Attached Disk" and so takes up 1 of the quota. I would suggest bringing the allocatable count down to 127 instead of modifying the test.

davidz627 commented:
Also when running the test on my machine, since the test `GET`'s `NodeInfo` the test fails if the `NodeInfo` object is not created yet since we could potentially install the driver right before and we don't wait for it to finish registering before proceeding with the test. This could lead to flakes in the future. Can be addressed separately. I was able to get around this by putting a sleep in the test before the line that `GET`'s the `NodeInfo` 🤦

jsafrane commented:
> test GET's NodeInfo the test fails if the NodeInfo object is not created yet since we could potentially install the driver right before and we don't wait for it to finish registering before proceeding with the test

I added wait.Poll around. It fails when a driver is in CSINodeInfo, but it has no limits set - we add the whole driver info with the limit set, right?

bertinatto commented:
/retest

jsafrane commented:
I'm hitting `context deadline exceeded` events and even `"I0802 12:16:01.098]               s: "unexpected error when reading response body. Please retry. Original error: local error: tls: bad record MAC"`, not sure what it means.

Anyway, let me know when GCE PD driver is fixed so we can continue with this PR.

jsafrane commented:
Filed https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/360

bertinatto commented:
/test pull-kubernetes-e2e-gce-csi-serial

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,280 @@
+			for _, pvc := range l.pvcs {
+				err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Delete(pvc.Name, nil)
+				framework.ExpectNoError(err, "failed cleaning up pvc")
+			}
```
If any of these is `err != nil`,  it won't try to delete the other objects?

jsafrane replied:
I reworked the cleanup completelly

jsafrane commented:
/retest
with new GCE driver

msau42 commented on test/e2e/storage/testsuites/testdriver.go:
```diff
@@ -147,6 +147,7 @@ const (
 	CapRWX                 Capability = "RWX"                 // support ReadWriteMany access modes
 	CapControllerExpansion Capability = "controllerExpansion" // support volume expansion for controller
 	CapNodeExpansion       Capability = "nodeExpansion"       // support volume expansion for node
+	CapVolumeLimits                   = "volumeLimits"        // support volume limits (applicable only to CSI drivers)
```
do we want to also support in-tree so that we can test migration?

davidz627 replied:
yeah, volume limits are also supported in-tree so testing both that and the migration from in-tree would be useful.

jsafrane replied:
In-tree drivers have their limits somewhere else. Is there a way how to detect a testdriver is CSI or not?

jsafrane replied:
Found it, `driverInfo.InTreePluginName`.

msau42 commented:
The PD driver manifest in testing-manifests needs to be updated first. @davidz627 can you take care of that?

davidz627 commented:
@jsafrane think you will have to update the CSI Driver version used in the tests explicitly to this: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/releases/tag/v0.5.2

image: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver:v0.5.2-gke.0

davidz627 commented:
/retest

jsafrane commented:
Now the test hits https://github.com/kubernetes-csi/external-attacher/pull/164. Volumes are marked as detached after `rateLimitExceeded` error. Please review https://github.com/kubernetes-csi/external-attacher/pull/165

```
I0806 02:21:26.147760       1 connection.go:180] GRPC call: /csi.v1.Controller/ControllerUnpublishVolume
I0806 02:21:26.147766       1 connection.go:181] GRPC request: {"node_id":"projects/gce-gci-upg-1-3-1-4-ctl-skew/zones/us-west1-b/instances/e2e-6d829a53ff-930d0-minion-group-63nj","volume_id":"projects/gce-gci-upg-1-3-1-4-ctl-skew/zones/us-west1-b/disks/pvc-edc8452d-189f-4237-b394-b8be2d754153"}
I0806 02:21:26.333412       1 request.go:530] Throttling request took 731.020924ms, request: PUT:https://10.0.0.1:443/apis/storage.k8s.io/v1beta1/volumeattachments/csi-44a03dbede65d12b35324b031b50b50f4bb630a1072c09258adb06b0370c03ff
I0806 02:21:26.335970       1 connection.go:183] GRPC response: {}
I0806 02:21:26.341190       1 connection.go:184] GRPC error: rpc error: code = Internal desc = unknown detach error: googleapi: Error 403: Too many pending operations on a resource., rateLimitExceeded
I0806 02:21:26.341212       1 csi_handler.go:394] Detached "csi-7a2dac28aa2ce05611788adfd189ac7215a9e27a0fc814f375576bd6e14dbbc4" with error rpc error: code = Internal desc = unknown detach error: googleapi: Error 403: Too many pending operations on a resource., rateLimitExceeded
```

jsafrane commented:
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce-device-plugin-gpu

jsafrane commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,280 @@
+			}
+			limit = int(*csiDriver.Allocatable.Count)
+			return true, nil
+		})
```
Fixed

jsafrane commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,280 @@
+		err = e2epod.WaitTimeoutForPodRunningInNamespace(l.cs, l.runningPod.Name, l.ns.Name, testSlowMultiplier*framework.PodStartTimeout)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Creating a pod with a volume over the limit")
```
Fixed

jsafrane commented:
So, I added support for in-tree volume limits. But then I needed to enable late binding by default for all(?) the tests, so I can get required amount of volumes usable by a single node.

I see it was already enabled so on GCE. Will anything break on the other clusters? I'm waiting for AWS results.

jsafrane commented:
Late binding breaks subpath tests and couple of others :-(

jsafrane commented:
> Late binding breaks subpath tests and couple of others :-(

It's broken on 1.14, 1.15 should be better.

jsafrane commented:
/test pull-kubernetes-bazel-build
/test pull-kubernetes-integration

jsafrane commented:
/hold
added commit with `canary` attacher to test with https://github.com/kubernetes-csi/external-attacher/pull/165

jsafrane commented:
/retest

jsafrane commented:
Tested on GCE, it takes 20 minutes (!) to create, attach, detach and delete 128 volumes (detach + delete are the slowest parts), both with in-tree and CSI drivers. I think it will time out the whole test suite frequently.

msau42 commented:
/assign @davidz627 
@jsafrane and I discussed a little bit and decided to disable this test by default until we can get the runtime down.  We can enable the tests in our external jobs where we have more room for longer tests.

davidz627 commented:
> @jsafrane and I discussed a little bit and decided to disable this test by default until we can get the runtime down. 
sgtm, we are working on this. 20mins for 128 disks going up then down is consistent with my observations. Probably some room for improvement here

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-bazel-test

bertinatto commented:
/test pull-kubernetes-e2e-gce-csi-serial

jsafrane committed:
Add e2e test for volume limits

The test creates as many PVs and pods as the driver/plugin reports to support on a
single node.

jsafrane commented:
/retest

davidz627 commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -1587,8 +1601,9 @@ func (a *awsDriver) GetDynamicProvisionStorageClass(config *testsuites.PerTestCo
+	delayedBinding := storagev1.VolumeBindingWaitForFirstConsumer
 
-	return testsuites.GetStorageClass(provisioner, parameters, nil, ns, suffix)
+	return testsuites.GetStorageClass(provisioner, parameters, &delayedBinding, ns, suffix)
```
do we still want to "turn on delayed binding" for these suites in this PR?

jsafrane replied:
No, we don't. Removed

davidz627 replied:
still here :)

davidz627 commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,335 @@
+	// And one extra pod with a CSI volume should get Pending with a condition
+	// that says it's unschedulable because of volume limit.
+	// BEWARE: the test may create lot of volumes and it's really slow.
+	ginkgo.It("should support volume limits [Slow][Serial]", func() {
```
would this test also be considered `[Disruptive]` as it would prevent further scheduling of pods using disks onto this node?

jsafrane replied:
IMO, the test does not break anything (e.g. it does not kill kubelet). It just needs to be alone on a node and that's why we have `[Serial]`.

davidz627 commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,335 @@
+			ginkgo.Skip(fmt.Sprintf("driver %s does not support volume limits", driverInfo.Name))
+		}
+		var dDriver DynamicPVTestDriver
+		if dDriver = driver.(DynamicPVTestDriver); dDriver == nil {
```
I thought these checks were all done in some helper function somewhere thats shared between test suites

jsafrane replied:
Technically the error path is not reachable. Retyping `driver` to `DynamicPVTestDriver` should always succeed, but I have a safety check just to be sure.

davidz627 commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,335 @@
+		defer l.resource.cleanupResource()
+
+		// Prepare cleanup
+		defer func() {
```
maybe split some of this cleanup function into helper functions, it's very long and complicated right now

jsafrane replied:
Moved to separate func.

davidz627 commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,335 @@
+
+		l.pvNames = sets.NewString()
+		ginkgo.By("Waiting for all PVCs to get Bound")
+		err = wait.Poll(5*time.Second, testSlowMultiplier*framework.PVBindingTimeout, func() (bool, error) {
```
is there an existing helper function for this?

jsafrane replied:
Not for array of PVCs, only single PVC.

jsafrane replied:
+ I need to collect PV names to make sure they're deleted before the test ends. It takes ages and the test suite could end before the PVs are detached + deleted.

jsafrane replied:
Moved to separate func.

davidz627 commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,335 @@
+		l.unschedulablePod, err = l.cs.CoreV1().Pods(l.ns.Name).Create(pod)
+
+		ginkgo.By("Waiting for the pod to get unschedulable")
+		err = wait.Poll(5*time.Second, framework.PodStartTimeout, func() (bool, error) {
```
helper function?

jsafrane replied:
Found `WaitForPodCondition`, but it's not that shorter.

davidz627 commented on test/e2e/storage/drivers/in_tree.go:
```diff
@@ -1472,8 +1482,9 @@ func (a *azureDriver) GetDynamicProvisionStorageClass(config *testsuites.PerTest
+	delayedBinding := storagev1.VolumeBindingWaitForFirstConsumer
 
-	return testsuites.GetStorageClass(provisioner, parameters, nil, ns, suffix)
+	return testsuites.GetStorageClass(provisioner, parameters, &delayedBinding, ns, suffix)
```
do we still want to turn on delayed binding?

davidz627 reviewed (commented):
lgtm modulo the extra code change for delayed binding

bertinatto commented:
/test pull-kubernetes-bazel-build

bertinatto commented:
@davidz627, @msau42: I addressed the comments, could you take another look?

davidz627 reviewed (commented):
couple small comments, lgtm besides those

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,346 @@
+		}
+
+		// Create the gigantic pod with all <limit> PVCs
+		ginkgo.By("Creating pod")
```
added

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,346 @@
+
+		// Create <limit> PVCs for one gigantic pod.
+		for i := 0; i < limit; i++ {
+			ginkgo.By(fmt.Sprintf("Creating volume %d/%d", i+1, limit))
```
moved above loop

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,346 @@
+		err = e2epod.WaitForPodCondition(l.cs, l.ns.Name, l.unschedulablePod.Name, "Unschedulable", framework.PodStartTimeout, func(pod *v1.Pod) (bool, error) {
+			if pod.Status.Phase == v1.PodPending {
+				for _, cond := range pod.Status.Conditions {
+					matched, _ := regexp.MatchString("max.+volume.+count", cond.Message)
```
[This is the string](https://github.com/bertinatto/kubernetes/blob/251b908c1affb84aa3dbed9ef9fdeafad59e0201/pkg/scheduler/algorithm/predicates/error.go#L59) it's trying to match, so it's only a space.

I think @jsafrane wanted to use the same approach used in the mock driver.

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -0,0 +1,346 @@
+		if len(nodeList.Items) != 0 {
+			nodeName = nodeList.Items[0].Name
+		} else {
+			e2elog.Failf("Unable to find ready and schedulable Node")
```
Replace one usage of `ginkgo.Fail` and removed a redundant `e2elog.Log`.

The rest seems consistent with other files (like `provisioning.go`), where (AFAIU) `ginkgo` is used to delimit/document the logic of the test and `e2elog` for logging additional information.

bertinatto replied:
Done

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-conformance-kind-ipv6

flaky job

bertinatto commented:
/test pull-kubernetes-bazel-build

bertinatto commented:
@davidz627, squashed commits and fixed some function calls (e.g., `MakeSecPod`) that have been moved recently. PTAL
